logstash-core/src/main/java/org/logstash/ackedqueue/FsUtil.java

```diff
@@ -0,0 +1,47 @@
+package org.logstash.ackedqueue;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * File System Utility Methods.
+ */
+public final class FsUtil {
+
+    private FsUtil() {
+    }
+
+    /**
+     * Checks if the request number of bytes of free disk space are available under the given
+     * path.
+     * @param path Directory to check
+     * @param size Bytes of free space requested
+     * @return True iff the
+     * @throws IOException on failure to determine free space for given path's partition
+     */
+    public static boolean hasFreeSpace(final String path, final long size) throws IOException {
+        final File[] partitions = File.listRoots();
+        File location = new File(path).getCanonicalFile();
+        boolean found = false;
+        while (!found) {
+            for (final File partition : partitions) {
+                if (partition.equals(location)) {
+                    found = true;
+                    break;
+                }
+            }
+            if (!found) {
+                location = location.getParentFile();
+                if (location == null) {
+                    throw new IllegalStateException(
```

Comment:
@colinsurprenant I think I explained this badly, the point is this:

If you input a valid directory and then go through its parents step by step and never arrive at the path of a partition, then you have some path that isn't part of any mounted partition.
At least on *nix this is impossible => `IllegalStateException` since I don't see how this could ever happen unless we have an actual error in the logic of this code => checked exception for a state that is assumed to be "impossible" seems wrong right?